sys/src/9/amd64/devvq.c

```diff
@@ -0,0 +1,638 @@
+/*
+ * This file is part of the Harvey operating system.  It is subject to the
+ * license terms of the GNU GPL v2 in LICENSE.gpl found in the top-level
+ * directory of this distribution and at http://www.gnu.org/licenses/gpl-2.0.txt
+ *
+ * No part of Harvey operating system, including this file, may be copied,
+ * modified, propagated, or distributed except according to the terms
+ * contained in the LICENSE.gpl file.
+ */
+
+// devvq.c ('#Q'): a generic virtqueue driver.
+
+#include	"u.h"
+#include	"../port/lib.h"
+#include	"mem.h"
+#include	"dat.h"
+#include	"fns.h"
+#include	"io.h"
+#include	"../port/error.h"
+
+// Include the definitions from VIRTIO spec v1.0
+// http://docs.oasis-open.org/virtio/virtio/v1.0/csprd02/listings/virtio_ring.h
+
+#include	"virtio_ring.h"
+
+#include	"virtio_config.h"
+#include	"virtio_pci.h"
+
+#include	"virtio_custom.h"
+
+#define TYPE(q)			((uint32_t)(q).path & 0x0F)
+#define DEV(q)			((uint32_t)(((q).path >> 4) & 0x0FFF))
+#define VQ(q)			((uint32_t)(((q).path >> 16) & 0x0FFFF))
+#define QID(c, t)		((((c) & 0x0FFF)<<4) | ((t) & 0x0F))
+#define VQQID(q, c, t)	((((q) & 0x0FFFF)<<16) | (((c) & 0x0FFF)<<4) | ((t) & 0x0F))
+
+#define BY2PG PGSZ
```

Comment:
is this about BY2PG? I think it is one remnant from 9front code which used this constant, can be changed.